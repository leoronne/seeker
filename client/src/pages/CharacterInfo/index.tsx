/* eslint-disable react/jsx-one-expression-per-line */
import React, { useEffect, useCallback, useState, useRef } from 'react';
import { useTranslation } from 'react-i18next';
import { Link, Redirect, useParams, useHistory } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import { useSnackbar } from 'notistack';
import { Formik } from 'formik';
import { CircularProgress, MenuItem } from '@material-ui/core';

import { LoaderSpinner } from '../../components';

import { characterResponse } from '../../__mocks__';
import { CharacterData } from '../../@types';
import { IState } from '../../store';

import api from '../../services/api';
import { editCharacter } from '../../store/modules/edittedCharacters/actions';

import { useStyles, FormControl, InputLabel, Select, TextField, ButtonOutlined } from '../../styles/MaterialUI';
import { Container, Header, Main, ReturnIcon, LinkIcon, TableInfo, Form, EditIcon, LikeIcon, CancelIcon } from './styles';

interface FormikProps {
  name: string;
  real_name: string;
  aliases: string;
  origin: string;
  gender: number;
  publisher: string;
  count_of_issue_appearances: string | number;
  birth: string;
}

const CharacterInfo: React.FC = () => {
  const { NODE_ENV } = process.env;

  const { t } = useTranslation();
  const { id } = NODE_ENV === 'test' ? { id: '4005-51713' } : useParams<{ id: string }>();
  const history = useHistory();
  const dispatch = useDispatch();
  const edittedCharacters = useSelector<IState, CharacterData[]>(state => state.edittedCharacters);
  const { enqueueSnackbar } = useSnackbar();
  const classes = useStyles();

  const [loading, setLoading] = useState(false);
  const [isFave, setIsFave] = useState(false);
  const [edit, setEdit] = useState(true);
  const [charData, setCharData] = useState<CharacterData>(NODE_ENV === 'test' ? characterResponse : ({} as CharacterData));

  const formRef = useRef(null);

  const INITIAL_VALUES: FormikProps = {
    name: charData?.name || '',
    real_name: charData?.real_name || '',
    aliases: charData?.aliases || '',
    origin: charData?.origin?.name || '',
    gender: charData?.gender || 0,
    publisher: charData?.publisher?.name || '',
    count_of_issue_appearances: charData?.count_of_issue_appearances || 0,
    birth: charData?.birth || '',
  };

  if (!id) {
    return <Redirect to="/" />;
  }

  const loadData = useCallback(async (charId: string) => {
    try {
      setLoading(true);
      const response = await api.get(`/characters/${charId}`);
      const { data } = response;

      if (!data) {
        history.push('/not-found');
      }

      const char_is_on_state = edittedCharacters.findIndex(favorite => favorite.id === data.id);

      if (char_is_on_state >= 0) {
        setCharData(edittedCharacters[char_is_on_state]);
        setIsFave(edittedCharacters[char_is_on_state]?.is_fave);
      } else setCharData(data);
    } catch (err) {
      enqueueSnackbar(err?.response?.data?.error ? err.response.data.error : err.message, { variant: 'error' });
      history.push('/');
    } finally {
      setLoading(false);
    }
  }, []);

  useEffect(() => {
    if (NODE_ENV !== 'test') loadData(id);
  }, [loadData, id]);

  const handleFave = useCallback(() => {
    dispatch(editCharacter(charData, !isFave));
    setIsFave(!isFave);
  }, [isFave, dispatch, charData]);

  const cancelEdit = useCallback(
    (setValue: (field: string, value: string | number | null, shouldValidate?: boolean | undefined) => void) => {
      setEdit(true);

      const keys = Object.keys(INITIAL_VALUES);
      const values = Object.values(INITIAL_VALUES);

      keys.forEach((key, index) => {
        setValue(key, values[index]);
      });
    },
    [charData]
  );

  const saveEdit = useCallback(
    (values: FormikProps) => {
      try {
        setEdit(true);
        const character: CharacterData = {
          ...charData,
          name: values.name,
          real_name: values.real_name,
          aliases: values.aliases,
          publisher: {
            api_detail_url: charData?.publisher?.api_detail_url || '',
            name: values.publisher,
            id: charData?.publisher?.id || 0,
          },
          gender: values.gender,
          origin: {
            api_detail_url: charData?.origin?.api_detail_url || '',
            name: values.origin,
            id: charData?.origin?.id || 0,
          },
          count_of_issue_appearances: values.count_of_issue_appearances,
          birth: values.birth,
        };
        setCharData(character);
        dispatch(editCharacter(character, isFave));

        enqueueSnackbar(t('success-edit'), { variant: 'success' });
      } catch (err) {
        enqueueSnackbar(err?.response?.data?.error ? err.response.data.error : err.message, { variant: 'error' });
      }
    },
    [charData, isFave]
  );

  if (loading) {
    return <LoaderSpinner color="#fff" />;
  }

  if (charData?.name) {
    document.title = `${charData?.name} - Seeker`;
  }

  return (
    <Container>
      <Header>
        <Link to="/">
          <ReturnIcon />
        </Link>

        <h1 data-testid="character-name">{charData?.name}</h1>

        <a href={charData?.site_detail_url} target="_blank" rel="noopener noreferrer" className="link-button">
          <LinkIcon />
        </a>
      </Header>

      <hr />
      <Main>
        <Formik
          initialValues={INITIAL_VALUES}
          onSubmit={async (values, { setSubmitting }) => {
            setSubmitting(true);
            saveEdit(values);
            setSubmitting(false);
          }}
        >
          {}
          {({ values, handleChange, handleBlur, handleSubmit, setFieldValue }) => (
            <Form onSubmit={handleSubmit} ref={formRef}>
              <div className="bio">
                {charData?.image?.original_url && <img src={charData?.image?.original_url} alt={charData?.name} />}

                {charData?.deck && <p>{charData?.deck}</p>}

                <div className="header-right">
                  {edit ? (
                    <div className="hover-button" onClick={() => setEdit(!edit)} data-testid="edit-button">
                      <EditIcon />
                    </div>
                  ) : (
                    <div className="hover-button" onClick={() => cancelEdit(setFieldValue)} data-testid="cancel-button">
                      <CancelIcon />
                    </div>
                  )}

                  <div className="hover-button" onClick={() => handleFave()} data-testid="favorite-button">
                    <LikeIcon fillcolor={isFave ? 'red' : 'white'} strokecolor={isFave ? 'red' : 'var(--color-primary)'} data-testid={`heart-svg-${isFave}`} />
                  </div>
                </div>
              </div>

              <TableInfo>
                <div className="field">
                  <TextField
                    disabled={edit}
                    name="name"
                    type="text"
                    variant="outlined"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.name}
                    label={t('super_name')}
                    inputProps={{ maxLength: 40, 'data-testid': 'name-input' }}
                  />
                </div>

                <div className="field">
                  <TextField
                    disabled={edit}
                    name="real_name"
                    type="text"
                    variant="outlined"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.real_name}
                    label={t('real_name')}
                  />
                </div>

                <div className="field">
                  <TextField
                    disabled={edit}
                    name="publisher"
                    type="text"
                    variant="outlined"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.publisher}
                    label={t('publisher')}
                  />
                </div>

                <div className="field">
                  <TextField
                    disabled={edit}
                    name="aliases"
                    type="text"
                    variant="outlined"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.aliases}
                    label={t('aliases')}
                  />
                </div>

                <div className="field">
                  <TextField disabled={edit} name="origin" type="text" variant="outlined" onChange={handleChange} onBlur={handleBlur} value={values.origin} label={t('origin')} />
                </div>

                <div className="field">
                  <FormControl variant="outlined">
                    <InputLabel>{t('gender')}</InputLabel>

                    <Select
                      name="gender"
                      disabled={edit}
                      value={values.gender}
                      onChange={e => setFieldValue('gender', Number(e.target.value))}
                      label={t('gender')}
                      MenuProps={{ classes: { paper: classes.selectOptions } }}
                    >
                      <MenuItem value={0}>{t(`gender-0`)}</MenuItem>
                      <MenuItem value={1}>{t(`gender-1`)}</MenuItem>
                      <MenuItem value={2}>{t(`gender-2`)}</MenuItem>
                    </Select>
                  </FormControl>
                </div>

                <div className="field">
                  <TextField disabled={edit} name="birth" type="text" variant="outlined" onChange={handleChange} onBlur={handleBlur} value={values.birth} label={t('birth')} />
                </div>

                <div className="field">
                  <TextField
                    disabled={edit}
                    name="count_of_issue_appearances"
                    type="text"
                    variant="outlined"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values.count_of_issue_appearances}
                    label={t('appearances')}
                  />
                </div>
              </TableInfo>

              {!edit && (
                <ButtonOutlined type={!loading ? 'submit' : 'button'} disabled={loading} data-testid="submit-button">
                  {loading ? (
                    <div>
                      <CircularProgress size={15} style={{ color: '#ccc' }} />
                    </div>
                  ) : (
                    <span>{t('save')}</span>
                  )}
                </ButtonOutlined>
              )}
            </Form>
          )}
        </Formik>
      </Main>
    </Container>
  );
};

export default CharacterInfo;
