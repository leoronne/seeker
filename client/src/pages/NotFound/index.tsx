import React from 'react';
import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';

import { Container, Content, HomeIcon, Left, Main, Right } from './styles';

const NotFound: React.FC = () => {
  const { t } = useTranslation();
  return (
    <Container data-testid="not-found-page">
      <Content>
        <Main>
          <Left>
            <h2>{t('error-page-1')}</h2>

            <p>{t('error-page-2')}</p>

            <div className="home-link">
              <Link to="/">
                <HomeIcon />
                <span>{t('return')}</span>
              </Link>
            </div>
          </Left>
          <Right />
        </Main>
      </Content>
    </Container>
  );
};

export default NotFound;
