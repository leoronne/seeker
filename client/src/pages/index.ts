export { default as Homepage } from './Homepage';
export { default as CharacterInfo } from './CharacterInfo';
export { default as NotFound } from './NotFound';
