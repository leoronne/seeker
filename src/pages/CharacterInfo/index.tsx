import React from 'react';

import { Container } from './styles';

const CharacterInfo: React.FC = () => {
  return (
    <Container>
      <h1>CharacterInfo</h1>
    </Container>
  );
};

export default CharacterInfo;
