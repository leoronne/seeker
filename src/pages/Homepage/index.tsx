import React from 'react';

import { Container } from './styles';

const Homepage: React.FC = () => {
  return (
    <Container>
      <h1>Homepage</h1>
    </Container>
  );
};

export default Homepage;
