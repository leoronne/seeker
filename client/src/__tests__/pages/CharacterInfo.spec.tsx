import React from 'react';
import { act, fireEvent, render, waitFor } from '@testing-library/react';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import notistack from 'notistack';

import { CharacterInfo } from '../../pages';

const mockedHistoryPush = jest.fn();
const mockedloadData = jest.fn();
const enqueueSnackbar = jest.fn();
const closeSnackbar = jest.fn();

jest.mock('react-router-dom', () => {
  return {
    useHistory: () => ({
      push: mockedHistoryPush,
    }),
    Link: ({ children }: { children: React.ReactNode }) => children,
  };
});

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key: string) => key }),
}));

jest.mock('notistack', () => ({
  useSnackbar: jest.fn(),
}));

jest.mock('../../hooks', () => {
  return {
    useLanguage: () => ({
      language: 'en',
    }),
  };
});

describe('CharacterInfo', () => {
  beforeEach(() => {
    mockedHistoryPush.mockClear();
    mockedloadData.mockClear();

    jest.spyOn(notistack, 'useSnackbar').mockImplementation(() => {
      return { enqueueSnackbar, closeSnackbar };
    });
  });

  const initialState = {};
  const mockStore = configureStore();
  const store = mockStore(initialState);

  it('should render', async () => {
    await act(async () => {
      const { asFragment } = render(
        <Provider store={store}>
          <CharacterInfo />
        </Provider>
      );

      const html = asFragment();

      expect(html).toMatchSnapshot();
    });
  });

  it('should click on edit button', async () => {
    await act(async () => {
      jest.mock('../../pages/CharacterInfo', () => ({
        loadData: mockedloadData,
      }));

      const { getByTestId, queryByTestId } = render(
        <Provider store={store}>
          <CharacterInfo />
        </Provider>
      );

      const editButton = await waitFor(() => getByTestId('edit-button'));

      fireEvent.click(editButton);

      await waitFor(() => {
        expect(queryByTestId('submit-button')).toBeTruthy();
      });
    });
  });

  it('should be able to click on the favorite button', async () => {
    await act(async () => {
      jest.mock('../../pages/CharacterInfo', () => ({
        loadData: mockedloadData,
      }));

      const { getByTestId, queryByTestId } = render(
        <Provider store={store}>
          <CharacterInfo />
        </Provider>
      );

      const favoriteButton = await waitFor(() => getByTestId('favorite-button'));

      fireEvent.click(favoriteButton);

      await waitFor(() => {
        expect(queryByTestId('heart-svg-true')).toBeTruthy();
      });
    });
  });

  it('should be able to cancel edit', async () => {
    await act(async () => {
      jest.mock('../../pages/CharacterInfo', () => ({
        loadData: mockedloadData,
      }));

      const { getByTestId } = render(
        <Provider store={store}>
          <CharacterInfo />
        </Provider>
      );

      const editButton = await waitFor(() => getByTestId('edit-button'));

      fireEvent.click(editButton);

      const nameInput = (await waitFor(() => getByTestId('name-input'))) as HTMLInputElement;

      const oldValue = nameInput.value;
      const newValue = 'Test Name';

      fireEvent.change(nameInput, { target: { value: newValue } });

      expect(nameInput.value).toBe(newValue);

      const cancelButton = await waitFor(() => getByTestId('cancel-button'));

      fireEvent.click(cancelButton);

      await waitFor(() => {
        expect(nameInput.value).toBe(oldValue);
      });
    });
  });

  it('should be able to edit character', async () => {
    await act(async () => {
      jest.mock('../../pages/CharacterInfo', () => ({
        loadData: mockedloadData,
      }));

      const { getByTestId } = render(
        <Provider store={store}>
          <CharacterInfo />
        </Provider>
      );

      const editButton = await waitFor(() => getByTestId('edit-button'));

      fireEvent.click(editButton);

      const nameInput = (await waitFor(() => getByTestId('name-input'))) as HTMLInputElement;

      const newValue = 'Test Name';

      fireEvent.change(nameInput, { target: { value: newValue } });

      expect(nameInput.value).toBe(newValue);

      const submitButton = await waitFor(() => getByTestId('submit-button'));

      fireEvent.click(submitButton);

      const nameH1 = await waitFor(() => getByTestId('character-name'));

      await waitFor(() => {
        expect(nameH1.textContent).toBe(newValue);
      });
    });
  });
});
