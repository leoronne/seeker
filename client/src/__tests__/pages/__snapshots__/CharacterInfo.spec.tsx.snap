// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterInfo should render 1`] = `
<DocumentFragment>
  <div
    class="sc-hHEjAm dbQeJu"
  >
    <div
      class="sc-dlMBXb UDFHJ"
    >
      <svg
        class="sc-hBMVcZ euuipo"
        fill="none"
        height="1em"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="19"
          x2="5"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 19 5 12 12 5"
        />
      </svg>
      <h1
        data-testid="character-name"
      >
        Ricky Landers
      </h1>
      <a
        class="link-button"
        href="https://comicvine.gamespot.com/ricky-landers/4005-51713/"
        rel="noopener noreferrer"
        target="_blank"
      >
        <svg
          class="sc-fnVYJo fGRxkB"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M1.5 13A1.5 1.5 0 003 14.5h8a1.5 1.5 0 001.5-1.5V9a.5.5 0 00-1 0v4a.5.5 0 01-.5.5H3a.5.5 0 01-.5-.5V5a.5.5 0 01.5-.5h4a.5.5 0 000-1H3A1.5 1.5 0 001.5 5v8zm7-11a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v5a.5.5 0 01-1 0V2.5H9a.5.5 0 01-.5-.5z"
            fill-rule="evenodd"
          />
          <path
            clip-rule="evenodd"
            d="M14.354 1.646a.5.5 0 010 .708l-8 8a.5.5 0 01-.708-.708l8-8a.5.5 0 01.708 0z"
            fill-rule="evenodd"
          />
        </svg>
      </a>
    </div>
    <hr />
    <main
      class="sc-kfYqjs cpacaf"
    >
      <form
        class="sc-bCwgka bRHBeO"
      >
        <div
          class="bio"
        >
          <img
            alt="Ricky Landers"
            src="https://comicvine1.cbsistatic.com/uploads/original/11113/111133571/4032871-ricky%20landers.jpg"
          />
          <p>
            Son of Gena Landers, and brother to Ray.
          </p>
          <div
            class="header-right"
          >
            <div
              class="hover-button"
              data-testid="edit-button"
            >
              <svg
                class="sc-fFSRdu joFIrn"
                fill="none"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M11 4H4a2 2 0 0 0-2 2v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2v-7"
                />
                <path
                  d="M18.5 2.5a2.121 2.121 0 0 1 3 3L12 15l-4 1 1-4 9.5-9.5z"
                />
              </svg>
            </div>
            <div
              class="hover-button"
              data-testid="favorite-button"
            >
              <svg
                class="sc-dIvqjp kWHRjR"
                data-testid="heart-svg-false"
                fill="none"
                fillcolor="white"
                height="1em"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                strokecolor="var(--color-primary)"
                viewBox="0 0 24 24"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M20.84 4.61a5.5 5.5 0 0 0-7.78 0L12 5.67l-1.06-1.06a5.5 5.5 0 0 0-7.78 7.78l1.06 1.06L12 21.23l7.78-7.78 1.06-1.06a5.5 5.5 0 0 0 0-7.78z"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="sc-fKgIGh gXKALV"
        >
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                data-shrink="true"
              >
                super_name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  data-testid="name-input"
                  disabled=""
                  maxlength="40"
                  name="name"
                  type="text"
                  value="Ricky Landers"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      super_name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                data-shrink="true"
              >
                real_name
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="real_name"
                  type="text"
                  value="Ricky Landers"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      real_name
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                data-shrink="true"
              >
                publisher
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="publisher"
                  type="text"
                  value="Marvel"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      publisher
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
                data-shrink="false"
              >
                aliases
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="aliases"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5"
                  >
                    <span>
                      aliases
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                data-shrink="true"
              >
                origin
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="origin"
                  type="text"
                  value="Human"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      origin
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root WithStyles(ForwardRef(FormControl))-root-7"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root WithStyles(ForwardRef(InputLabel))-root-8 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined WithStyles(ForwardRef(InputLabel))-outlined-10 MuiFormLabel-filled"
                data-shrink="true"
              >
                gender
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <div
                  aria-disabled="true"
                  aria-haspopup="listbox"
                  aria-labelledby="mui-component-select-gender"
                  class="MuiSelect-root WithStyles(ForwardRef(Select))-root-11 MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled Mui-disabled"
                  id="mui-component-select-gender"
                  role="button"
                >
                  gender-1
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput"
                  name="gender"
                  tabindex="-1"
                  value="1"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined Mui-disabled"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      gender
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-disabled Mui-disabled"
                data-shrink="false"
              >
                birth
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="birth"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5"
                  >
                    <span>
                      birth
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
          <div
            class="field"
          >
            <div
              class="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-2"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-disabled Mui-disabled MuiFormLabel-filled"
                data-shrink="true"
              >
                appearances
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root Mui-disabled Mui-disabled MuiInputBase-formControl"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled Mui-disabled"
                  disabled=""
                  name="count_of_issue_appearances"
                  type="text"
                  value="9"
                />
                <fieldset
                  aria-hidden="true"
                  class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
                  >
                    <span>
                      appearances
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
      </form>
    </main>
  </div>
</DocumentFragment>
`;
